src/Npgsql/TypeHandlers/TextHandler.cs

```diff
@@ -168,6 +168,45 @@ ValueTask<ArraySegment<char>> INpgsqlTypeHandler<ArraySegment<char>>.Read(Npgsql
             throw new NpgsqlSafeReadException(new NotSupportedException("Only writing ArraySegment<char> to PostgreSQL text is supported, no reading."));
         }
 
+        async ValueTask<byte[]> INpgsqlTypeHandler<byte[]>.Read(NpgsqlReadBuffer buf, int byteLen, bool async, FieldDescription fieldDescription)
+        {
+            var bytes = new byte[byteLen];
+            if (buf.ReadBytesLeft >= byteLen)
+            {
+                buf.ReadBytes(bytes, 0, byteLen);
+                return bytes;
+            }
+
+            if (byteLen <= buf.Size)
+            {
+                // The bytes can fit in our read buffer, read it.
+                while (buf.ReadBytesLeft < byteLen)
+                    await buf.ReadMore(async);
+                buf.ReadBytes(bytes, 0, byteLen);
+                return bytes;
+            }
+
+            // Bad case: the bytes don't fit in our buffer.
+            // This is rare - will only happen in CommandBehavior.Sequential mode (otherwise the
+            // entire row is in memory). Tweaking the buffer length via the connection string can
+            // help avoid this.
+
+            var pos = 0;
+            while (true)
+            {
+                var len = Math.Min(buf.ReadBytesLeft, byteLen - pos);
+                buf.ReadBytes(bytes, pos, len);
+                pos += len;
+                if (pos < byteLen)
+                {
+                    await buf.ReadMore(async);
+                    continue;
+                }
+                break;
+            }
+            return bytes;
```

Comment:
The last thing which should be done before merge is splitting the `Read` method (forgot to mention that). It would improve performance. The code should looks like:

```csharp
async ValueTask<byte[]> INpgsqlTypeHandler<byte[]>.Read(NpgsqlReadBuffer buf, int byteLen, bool async, FieldDescription fieldDescription)
{
    var bytes = new byte[byteLen];
    if (buf.ReadBytesLeft >= byteLen)
    {
        buf.ReadBytes(bytes, 0, byteLen); 
        return bytes;
    }

    return ReadLong();

    // it's a local function which uses an async state machine
    async ValueTask<byte[]> ReadLong()
    {
        // code from else
    }
}
```

You also can change the condition and call `ReadLong` on it, but it`s up to you.

```csharp
var bytes = new byte[byteLen];
if (buf.ReadBytesLeft < byteLen)
    return ReadLong();

buf.ReadBytes(bytes, 0, byteLen); 
return bytes;
```

For example see c67a1f4fe4fd935a2606f6166bb7a64fde866331.